google/cloud/storage/internal/curl_client.cc

```diff
@@ -1056,6 +1068,73 @@ CurlClient::WriteObjectXml(InsertObjectStreamingRequest const& request) {
       std::unique_ptr<internal::ObjectWriteStreambuf>(std::move(buf)));
 }
 
+std::pair<Status, ObjectMetadata> CurlClient::InsertObjectMediaMultipart(
+    InsertObjectMediaRequest const& request) {
+  CurlRequestBuilder builder(
+      upload_endpoint_ + "/b/" + request.bucket_name() + "/o", upload_factory_);
+  SetupBuilder(builder, request, "POST");
+  // Pick a separator that does not conflict with the request contents:
+  auto boundary = PickBoundary(request.contents());
+  builder.AddHeader("content-type: multipart/related; boundary=" + boundary);
+  builder.AddQueryParameter("uploadType", "multipart");
+  builder.AddQueryParameter("name", request.object_name());
+
+  std::unique_ptr<internal::CurlStreambuf> buf(
+      new internal::CurlStreambuf(builder.BuildUpload(),
+                                  client_options().upload_buffer_size()));
+  ObjectWriteStream writer(std::move(buf));
+
+  nl::json metadata = nl::json::object();
+  if (request.HasOption<MD5HashValue>()) {
+    metadata["md5Hash"] = request.GetOption<MD5HashValue>().value();
+  }
+
+  std::string marker = "--" + boundary + "\r\n";
+  writer << marker
+         << "content-type: application/json; charset=UTF-8\r\n\r\n"
+         << metadata.dump() << "\r\n"
+         << marker;
+
+  if (not request.HasOption<ContentType>()) {
+    writer << "content-type: application/octet-stream\r\n";
+  } else {
+    writer << "content-type: " << request.GetOption<ContentType>().value() << "\r\n";
+  }
+  writer << "\r\n" << request.contents() << "\r\n--" << boundary << "--\r\n";
```

Comment:
Nit: Given the comment above, write the end here as `marker << "--"` rather than writing out `"--" << boundary << "--"` again.

Reply:
Done.